openedx/core/djangoapps/safe_sessions/middleware.py

```diff
@@ -768,6 +768,80 @@ def _get_encrypted_request_headers(request):
+        to match the new email.
+    2) It will invalidate any future session on other browsers where
+        the user's email does not match its session email.
+
+    This middleware ensures that the sessions in other browsers
+    are invalidated when a user changes their email in one browser.
+    The active session in which the email change is made will remain valid.
+
+    The user's email is stored in their session and JWT cookies during login
+    and gets updated when the user changes their email.
+    This middleware checks for any mismatch between the stored email
+    and the current user's email in each request, and if found,
+    it invalidates/flushes the session and mark cookies for deletion in request
+    which are then deleted in the process_response of SafeSessionMiddleware.
+    """
+
+    def process_request(self, request):
+        """
+        Invalidate the user session if there's a mismatch
+        between the email in the user's session and request.user.email.
+        """
+        if request.user.is_authenticated:
+            user_session_email = request.session.get('email', None)
+            are_emails_mismatched = user_session_email is not None and request.user.email != user_session_email
+            EmailChangeMiddleware._set_session_email_match_custom_attributes(are_emails_mismatched)
+            if settings.ENFORCE_SESSION_EMAIL_MATCH and are_emails_mismatched:
+                # Flush the session and mark cookies for deletion.
+                request.session.flush()
+                request.user = AnonymousUser()
+                _mark_cookie_for_deletion(request)
+
+    def process_response(self, request, response):
+        """
+        Update the logged-in cookies if the email change was requested
+        """
+        if request.user.is_authenticated and request_cache.get_cached_response('email_change_requested').is_found:
+            # Update the JWT cookies with new user email
+            response = set_logged_in_cookies(request, response, request.user)
+
+        return response
+
+    @staticmethod
+    def register_email_change(request, email):
+        """
+        Stores the fact that an email change happened.
+
+        1. Sets the email in session for later comparison.
+        2. Sets a request level variable to mark that the user email change was requested.
+        """
+        request.session['email'] = email
+        request_cache.set('email_change_requested', True)
+
+    @staticmethod
+    def _set_session_email_match_custom_attributes(are_emails_mismatched):
+        """
+        Sets custom attributes of session_email_match
+        """
+        # .. custom_attribute_name: session_email_match
+        # .. custom_attribute_description: Indicates whether there is a match between the
+        #      email in the user's session and the current user's email in the request.
+        set_custom_attribute('session_email_match', not are_emails_mismatched)
```

Comment:
> Do you have a ticket for removing ENFORCE_SESSION_EMAIL_MATCH after rollout?

We will create ticket in our backlog for this

> We may want to remove both these custom attributes after rollout, unless we only set this attribute for mismatches.
> Also, this could be set_custom_attribute('session_email_mismatch', are_emails_mismatched). After rollout, we could move it under the if are_emails_mismatched: conditional.

Yes, we can have this attribute under the  ` if are_emails_mismatched: ` conditional after rollout.

Reply:
> Yes, we can have this attribute under the if are_emails_mismatched: conditional after rollout.
Maybe add this to the ticket.

Also, if you haven't already, maybe add a reference to the ticket in the toggle annotation.

Reply:
Marking as resolved.